addons/road-generator/road_segment.gd

```diff
@@ -227,12 +227,17 @@ func _update_curve():
 	curve.set_point_tilt(1, end_point.rotation.z)
 
 
-func _normal_for_offset(curve:Curve3D, offset:float):
-	var point1 = curve.interpolate_baked(offset - 0.001) # avoid below 0
-	var point2 = curve.interpolate_baked(offset + 0.001) # avoid over maxlen
-	var uptilt = curve.interpolate_baked_up_vector(offset, true)
-	var tangent:Vector3 = (point2 - point1)
-	return uptilt.cross(tangent).normalized()
+func _normal_for_offset(curve: Curve3D, offset_s: float, loop: float, loops: float):
```

Comment:
Could we add a docstring for what each of these args really represent?
loop vs loops feels a bit confusing to me. Maybe should be loop_idx and loop_count (to reference the first is an index, the second being the total number)? Also adding a return annotation could help with clarity too, ie Vector3 (does it need to be normalized??)

Solution does feel cleaner, not having to due baked offsets.

Reply:
Initially, those variables had completely different names. But, I aligned them with the variables you had incorporated into the _build_geo and _insert_geo_loop routines. If someone understands the usage in those routines, then they should also understand the usage in _normal_for_offset.

At first glance, those terms are not overly intuitive. But, for anyone who is familiar with a term used in 3D modeling such as "edge loop", the terms "loop" and "loops" should make a lot of sense.

I'll try to abstract something useful into a docstring. Regarding the function's return value: It is a "basis" value. Hence, it is inherently normalized.

Reply:
The comment came out of a convention we came up with in my other eng team, which was more about avoiding a situation of having two variables with only one letter difference (especially when the difference is a plural). I won't die on this hill though, especially if it's a parallel to other parts of the codebase in the meantime.

And gotcha on basis -  I was thinking that sometimes you still needed to manually run it (and sometimes a source of errors over time), but if you're understanding is that the basis itself is defined to always be normalized, then I'm happy to trust that.

Reply:
Added a docstring to _normal_for_offset in road_segment.gd. But, noted that there were some unnecessary extraneous inputs. Re-organized the inputs and did away with "loop" and "loops" since another input variable already supplied the necessary information.

Reply:
Regarding "basis" always being normalized, I think it would be best to say that may not always be true. But, in this case, the basis IS always a value between 0 and 1. You may also find that, frequently, when basis is consumed, it must be supplied in a normalized state.